extensions/amp-form/0.1/amp-form-textarea.js

```diff
@@ -0,0 +1,239 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {Services} from '../../../src/services';
+import {computedStyle, px, setStyle} from '../../../src/style';
+import {dev, devAssert} from '../../../src/log';
+import {iterateCursor, removeElement} from '../../../src/dom';
+import {listen, listenOncePromise} from '../../../src/event-helper';
+import {throttle} from '../../../src/utils/rate-limit';
+
+const AMP_FORM_TEXTAREA_EXPAND_ATTR = 'autoexpand';
+
+const AMP_FORM_TEXTAREA_SHRINK_DISABLED_ATTR = 'autoshrink-disabled';
+
+const MIN_EVENT_INTERVAL_MS = 100;
+
+const AMP_FORM_TEXTAREA_CLONE_CSS = 'i-amphtml-textarea-clone';
+
+const AMP_FORM_TEXTAREA_MAX_CSS = 'i-amphtml-textarea-max';
+
+const AMP_FORM_TEXTAREA_HAS_EXPANDED_DATA = 'iAmphtmlHasExpanded';
+
+class AmpFormWithTextarea {
+  /**
+   * @param {!Element} form
+   */
+  constructor(form) {
+    /** @private */
+    this.win_ = devAssert(form.ownerDocument.defaultView);
+
+    /** @private */
+    this.unlisteners_ = [];
+
+    this.unlisteners_.push(listen(form, 'input', e => {
+      const element = dev().assertElement(e.target);
+      if (element.tagName != 'TEXTAREA' ||
+          !element.hasAttribute(AMP_FORM_TEXTAREA_EXPAND_ATTR)) {
+        return;
+      }
+
+      maybeResizeTextarea(element);
+    }));
+
+    this.unlisteners_.push(listen(form, 'mousedown', e => {
+      if (e.which != 1) {
+        return;
+      }
+
+      const element = dev().assertElement(e.target);
+      if (element.tagName != 'TEXTAREA' ||
+          !element.hasAttribute(AMP_FORM_TEXTAREA_EXPAND_ATTR)) {
+        return;
+      }
+
+      maybeRemoveResizeBehavior(element);
+    }));
+
+    const throttledResize = throttle(this.win_, () => {
+      resizeFormTextareaElements(form);
+    }, MIN_EVENT_INTERVAL_MS);
+    this.unlisteners_.push(listen(this.win_, 'resize', throttledResize));
+  }
+
+  /**
+   * Cleanup any consumed resources
+   */
+  dispose() {
+    this.unlisteners_.forEach(unlistener => unlistener());
+  }
+}
+
+/**
+ * Install expandable textarea behavior for the given form.
+ *
+ * This method can be removed when browsers implement `height: max-content`
+ * for the textarea element.
+ * https://github.com/w3c/csswg-drafts/issues/2141
+ * @param {!Element} form
+ */
+export function installAmpFormTextarea(form) {
+  new AmpFormWithTextarea(form);
+}
+
+/**
+ * Attempt to resize all textarea elements within the given form.
+ * @param {!Element} form
+ */
+function resizeFormTextareaElements(form) {
+  iterateCursor(form.getElementsByTagName('textarea'), element => {
```

Comment:
`HTMLFormElement.elements` to get non-descendant inputs.

Reply:
If we aren't handling it on an `O(#forms)` basis, then I think I'll use `rootNode.querySelectorAll` to query for the elements